build/url_test.go

```diff
@@ -0,0 +1,174 @@
+	"https://github.com/cockroachdb/cockroach/commits/%s",
+	"https://github.com/cockroachdb/cockroach/issues/%d",
+	// These are AWS endpoints.
+	"http://metadata/",
+	"http://instance-data/latest/meta-data/public-ipv4",
+	// All these are auto-generated and thus valid by definition.
+	"https://registry.yarnpkg.com/",
+	// XML namespace identifiers, these are not really HTTP endpoints.
+	"https://www.w3.org/",
+	"http://www.bohemiancoding.com/sketch/ns",
+	"http://www.w3.org/",
+	// These report 404 for non-API GETs.
+	"http://ignored:ignored@errors.cockroachdb.com/",
+	"https://cockroachdb.github.io/distsqlplan/",
+	"https://ignored:ignored@errors.cockroachdb.com/",
+	"https://register.cockroachdb.com",
+	"https://www.googleapis.com/auth",
+	// GitHub reports 404 for private repos without auth
+	"https://github.com/cockroachlabs/registration/",
+	// Regexp not too smart about HTML...
+	`http://www.cockroachlabs.com">http://www.cockroachlabs.com</a></span`,
+}
+
+func testURLLiveness(t *testing.T, pkg *build.Package) {
+	t.Parallel()
+
+	cmd, stderr, filter, err := dirCmd(pkg.Dir, "git", "grep", "-nE", urlRE)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := cmd.Start(); err != nil {
+		t.Fatal(err)
+	}
+
+	uniqueURLs := map[string][]string{}
+
+	if err := stream.ForEach(filter, func(s string) {
+		for _, match := range re.FindAllString(s, -1) {
+			// Remove punctuation after the URL
+			match = strings.TrimRight(match, ".,;\\\">`]")
+			// Remove the HTML target
+			n := strings.LastIndex(match, "#")
+			if n != -1 {
+				match = match[:n]
+			}
+			// Add the source reference to the list
+			l, _ := uniqueURLs[match]
+			l = append(l, s)
+			uniqueURLs[match] = l
+		}
+	}); err != nil {
+		t.Error(err)
+	}
+
+	if err := cmd.Wait(); err != nil {
+		if out := stderr.String(); len(out) > 0 {
+			t.Fatalf("err=%s, stderr=%s", err, out)
+		}
+	}
+
+	var wg sync.WaitGroup
+	var sem = make(chan struct{}, concurrentRequests)
+
+	// This test doesn't care that https certificates are invalid.
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+	client := &http.Client{Transport: tr}
+
+	for url, locs := range uniqueURLs {
+		ignore := false
+		for _, i := range ignored {
+			if strings.HasPrefix(url, i) {
+				ignore = true
+				break
+			}
+		}
+		if ignore {
+			continue
+		}
+
+		sem <- struct{}{}
+		wg.Add(1)
+		go func(url string, locs []string) {
+			defer func() { wg.Done(); <-sem }()
+
+			fmt.Fprintf(os.Stderr, "Checking %s...\n", url)
+			var buf bytes.Buffer
+			fmt.Fprintln(&buf, "Found in:")
+			for _, loc := range locs {
+				fmt.Fprintln(&buf, loc)
+			}
+
+			resp, err := client.Head(url)
+			if err != nil {
+				t.Errorf("while accessing %s: %v\n%s", url, err, buf.String())
+				return
+			}
+			if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+				if resp.StatusCode == 405 {
+					// The server doesn't like HEAD. Try GET.
+					resp, err = client.Get(url)
```

Comment:
I think you need to fully read and then Close resp.Body to make the client happy, otherwise the underlying connection isn't freed for reuse.

Copy(ioutil.Discard, resp.Body) && resp.Body.Close if memory serves